libraries/chain/block_header_state.cpp

```diff
@@ -7,10 +7,62 @@
 
 namespace eosio::chain {
 
-// moved this warning out of header so it only uses once
-#warning TDDO https://github.com/AntelopeIO/leap/issues/2080
-// digest_type           compute_finalizer_digest() const { return id; };
+// this is a versioning scheme that is separate from protocol features that only
+// gets updated if a protocol feature causes a breaking change to light block
+// header validation
+
+// data for finality_digest
+struct finality_digest_data_v0_t {
+   uint32_t    major_version{light_header_protocol_version_major};
+   uint32_t    minor_version{light_header_protocol_version_minor};
+   uint32_t    active_finalizer_policy_generation {0};
+   digest_type finality_tree_digest;
+   digest_type base_and_active_finalizer_policy_digest;
+};
+
+// data for base_and_active_finalizer_policy_digest
+struct base_and_active_finalizer_policy_digest_data_t {
+   digest_type active_finalizer_policy_digest;
+   digest_type base_digest;
+};
+
+// data for base_digest
+struct base_digest_data_t {
+   block_header                             header;
+   finality_core                            core;
+   flat_map<uint32_t, finalizer_policy_ptr> finalizer_policies;
+   proposer_policy_ptr                      active_proposer_policy;
+   flat_map<block_timestamp_type, proposer_policy_ptr>  proposer_policies;
+   protocol_feature_activation_set_ptr      activated_protocol_features;
+};
+
+digest_type block_header_state::compute_finalizer_digest() const {
+   auto active_finalizer_policy_digest = fc::sha256::hash(*active_finalizer_policy);
+
+   base_digest_data_t base_digest_data {
+      .header                      = header,
+      .core                        = core,
+      .finalizer_policies          = finalizer_policies,
+      .active_proposer_policy      = active_proposer_policy,
+      .proposer_policies           = proposer_policies,
+      .activated_protocol_features = activated_protocol_features
+   };
+   auto base_digest = fc::sha256::hash(base_digest_data);
```

Comment:
Thanks for your comments about the two pointers. Will correct them.

I used structs here so that all the data fields to be hashed are clearly spelled out.

By "explicitly serializing", do you mean using `fc:sha256`'s `enc` and `fc::raw::pack`?

Reply:
Done.